chia/data_layer/data_layer.py

```diff
@@ -441,22 +451,33 @@ async def upload_files(self, tree_id: bytes32) -> None:
         await self._update_confirmation_status(tree_id=tree_id)
 
         root = await self.data_store.get_tree_root(tree_id=tree_id)
+        latest_generation = root.generation
+        # Don't store full tree files before this generation.
+        full_tree_first_publish_generation = max(0, latest_generation - self.full_file_storage_window + 1)
         publish_generation = min(singleton_record.generation, 0 if root is None else root.generation)
         # If we make some batch updates, which get confirmed to the chain, we need to create the files.
         # We iterate back and write the missing files, until we find the files already written.
         root = await self.data_store.get_tree_root(tree_id=tree_id, generation=publish_generation)
         while publish_generation > 0:
-            write_file_result = await write_files_for_root(self.data_store, tree_id, root, self.server_files_location)
+            write_file_result = await write_files_for_root(
+                self.data_store,
+                tree_id,
+                root,
+                self.server_files_location,
+                full_tree_first_publish_generation,
+            )
             if not write_file_result.result:
                 # this particular return only happens if the files already exist, no need to log anything
                 break
             try:
                 if uploaders is not None and len(uploaders) > 0:
                     request_json = {
                         "store_id": tree_id.hex(),
-                        "full_tree_filename": write_file_result.full_tree.name,
                         "diff_filename": write_file_result.diff_tree.name,
                     }
+                    if write_file_result.full_tree is not None:
+                        request_json["full_tree_filename"] = write_file_result.full_tree.name
```

Comment:
https://github.com/Chia-Network/chia-blockchain/blob/2126598941f9fab021929e58a2ec9fb276f1f547/chia/data_layer/s3_plugin_service.py#L145

The plugin requires that the `full_tree_filename` key be present and it assumes the value will be a string.  Seems like we're looking at a breaking API change here.  I'll ask around about this.

This also seems to imply we are not testing this case.

Reply:
I created PR https://github.com/Chia-Network/chia-blockchain/pull/16135 to update the plugin.
still nothing tested in the plugin though.

Reply:
Florin made some small updates to the plugin to not need the full file name